src/System.Memory/src/System/SpanExtensions.cs

```diff
@@ -274,5 +274,104 @@ public static void CopyTo<T>(this T[] array, Span<T> destination)
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool Overlaps<T>(this Span<T> first, ReadOnlySpan<T> second)
+        {
+            return Overlaps((ReadOnlySpan<T>)first, second);
+        }
+
+        /// <summary>
+        /// Determines whether two sequences overlap and outputs the element offset.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool Overlaps<T>(this Span<T> first, ReadOnlySpan<T> second, out int elementOffset)
+        {
+            return Overlaps((ReadOnlySpan<T>)first, second, out elementOffset);
+        }
+
+        /// <summary>
+        /// Determines whether two sequences overlap.
+        /// </summary>
+        public static bool Overlaps<T>(this ReadOnlySpan<T> first, ReadOnlySpan<T> second)
+        {
+            // note: the calculations below overflow if second.IsEmpty;
+            //       the case second.IsEmpty MUST be handled separately
+            if (first.IsEmpty || second.IsEmpty)
+            {
+                return false;
+            }
+
+            IntPtr byteOffset = Unsafe.ByteOffset(
+                ref first.DangerousGetPinnableReference(),
+                ref second.DangerousGetPinnableReference());
+
+            if (Unsafe.SizeOf<IntPtr>() == sizeof(int))
+            {
+                return (uint)byteOffset < (uint)(first.Length * Unsafe.SizeOf<T>()) ||
+                       (uint)byteOffset > (uint)-(second.Length * Unsafe.SizeOf<T>());
+            }
+            else
+            {
+                return (ulong)byteOffset < (ulong)((long)first.Length * Unsafe.SizeOf<T>()) ||
+                       (ulong)byteOffset > (ulong)-((long)second.Length * Unsafe.SizeOf<T>());
+            }
+        }
+
+        /// <summary>
+        /// Determines whether two sequences overlap and outputs the element offset.
+        /// </summary>
+        public static bool Overlaps<T>(this ReadOnlySpan<T> first, ReadOnlySpan<T> second, out int elementOffset)
+        {
+            // note: the calculations below overflow if second.IsEmpty;
+            //       the case second.IsEmpty MUST be handled separately
+            if (first.IsEmpty || second.IsEmpty)
+            {
+                elementOffset = 0;
+                return false;
+            }
+
+            IntPtr byteOffset = Unsafe.ByteOffset(
+                ref first.DangerousGetPinnableReference(),
+                ref second.DangerousGetPinnableReference());
+
+            if (Unsafe.SizeOf<IntPtr>() == sizeof(int))
+            {
+                if ((uint)byteOffset < (uint)(first.Length * Unsafe.SizeOf<T>()) ||
+                    (uint)byteOffset > (uint)-(second.Length * Unsafe.SizeOf<T>()))
+                {
+                    if ((int)byteOffset % Unsafe.SizeOf<T>() != 0)
+                        ThrowHelper.ThrowArgumentException_OverlapAlignmentMismatch();
```

Comment:
We had a discussion about it at the API review, but I now wonder if we really do want to throw here. A reusable library, e.g. a library that transforms between two buffers passed in by the app, does not have a continent way to check for this condition and so will have to either try/catch when calling Overlaps or let the exception propagate.

If we say that the API is for transformation libraries to tell whether data can end up being overridden, then maybe instead of throwing we round the rational number up to the next whole number? i.e. we return the number of elements that would overlap even if only partially.

Reply:
The exception only ever throws if some code crafts spans from pointers in an `unsafe` block or uses `NonPortableCast` from a smaller type to a larger type. I guess this code also would also be responsible for catching the exception, not any method that the spans are passed to.

Since the condition is so narrow and only happens with `unsafe` and `NonPortable` code, I would probably just ignore this corner case. Users of `unsafe` and `NonPortable` code must be aware that things can go wrong.

Reply:
@KrzysztofCwalina verdict?

Reply:
> library that transforms between two buffers passed in by the app

It may be useful to try to write a example that transforms between two buffers passed in by the app, and uses Overlaps to deal with overlapping buffers.

Reply:
@jkotas There's an example in https://github.com/dotnet/corefx/pull/24980#issuecomment-341466838 above (`RepeatEachElement`). Do you mean to include it in the unit tests?

Reply:
I meant it to help us decide on the right behavior for elementOffset (throw, align up, align down, ...). I do not see `RepeatEachElement` example that uses the elementOffset overload. Is there one?

Reply:
@jkotas 


```csharp
static void NegateEachElement(ReadOnlySpan<int> source, Span<int> destination)
{
    if (source.Length != destination.Length)
        throw new ArgumentException();

    if (source.Overlaps1(destination, out int elementOffset) && elementOffset > 0)
        source = source.ToArray();

    for (int i = 0; i < source.Length; i++)
        destination[i] = -source[i];
}

[Fact]
public static void TestAligned()
{
    for (int i = 0; i < 128 - 12; i++)
    {
        int[] a = Enumerable.Range(0, 128).ToArray();

        ReadOnlySpan<int> source = a.AsReadOnlySpan().Slice(32, 12);

        Span<int> expected = new int[a.Length].AsSpan().Slice(i, 12);
        Span<int> actual = a.AsSpan().Slice(i, 12);

        NegateEachElement(source, expected);
        NegateEachElement(source, actual);

        Assert.Equal(expected.ToArray(), actual.ToArray());
    }
}

[Fact]
public static void TestUnaligned()
{
    for (int i = 0; i < 128 - 12; i++)
    {
        int[] a = Enumerable.Range(0, 128).ToArray();

        ReadOnlySpan<int> source = a.AsReadOnlySpan().AsBytes()
            .Slice(16 * sizeof(int) + 2, 12 * sizeof(int))
            .NonPortableCast<byte, int>();

        Span<int> expected = new int[a.Length].AsSpan().Slice(i, 12);
        Span<int> actual = a.AsSpan().Slice(i, 12);

        NegateEachElement(source, expected);
        NegateEachElement(source, actual);

        Assert.Equal(expected.ToArray(), actual.ToArray());
    }
}
```

`TestUnaligned` passes with rounding away from zero. It fails with rounding towards 0.

Reply:
@jkotas https://gist.github.com/ektrah/8eea718c7abee180490776cde632396f

The tests pass with rounding away from zero. The unaligned tests fail with rounding towards zero, +inf, or -inf.

Reply:
After a bit more experimentation, I take my comment above (ignore this corner case) back. I like rounding away from zero to the next multiple of sizeof(T). I think it solves the problem quite elegantly.

Shall I go ahead and update the code?

Reply:
I think so.